### ISSUE TITLE
Value.Builder support in Struct.Builder APIs.

### ISSUE THREAD

**Issue by tuckerje** [protocolbuffers/protobuf#14343]
**What language does this apply to?**  Java
If it's a proto syntax change, is it for proto2 or proto3? Possibly?
If it's about generated code change, what programming language? Java 

**Describe the problem you are trying to solve.**
Right now, the Struct.Builder APIs don't support Value.Builder as inputs, only Value. It requires deferred computation for the struct if the value is not set on the Value.Builder ahead of type.  

**Describe the solution you'd like**
If Struct.Builder APIs (primarily, putFields API) can accept a builder as input, the reference back to the Value.Builder.build() could be called from Struct.Builder.build() and deferrred computation would not be needed.

**Describe alternatives you've considered**
Deferred computation with lambdas has already been considered. However, the order of build calls needs to be maintained and this limits opportunities to execute logic in parallel. For instance, if I build a recursive Struct structure like a tree, I must build from leaf to root. So, if traversing a tree structure from root to leaf, a second traversal is required to build. 

**Additional context**
N/A
